### ISSUE TITLE
False Positive: IWYU suggests to remove <iterator> and <algorithm> on template function

### ISSUE THREAD

**Issue by carlosgalvezp** [include-what-you-use/include-what-you-use#1293]
Hi!

We encounter a FP on IWYU trunk. Consider the following minimal example:

```cpp
#include <algorithm>
#include <array>
#include <cstddef>
#include <iterator>

template <std::size_t size>
void foo(std::array<float, size> const& input)
{
    float output[size]{};
    std::copy(std::begin(input), std::end(input), std::begin(output));
}

```

Running on IWYU with default flags gives:

```
$ ./include-what-you-use /tmp/foo.cpp 

/tmp/foo.cpp should add these lines:

/tmp/foo.cpp should remove these lines:
- #include <algorithm>  // lines 1-1
- #include <iterator>  // lines 4-4

The full include-list for /tmp/foo.cpp:
#include <array>    // for array
#include <cstddef>  // for size_t
---
```

However it's not correct:
- `<algorithm>` is needed for `std::copy`.
- `<iterator>` is needed for `std::begin`.

The problem goes away if I make the function non-templated.

**Comment by kimgr** [comment]
with `-Xiwyu -v7`:
```
tests/1293/t.cc:10:5: (4) [ CallExpr ] 0x1283477d8 CallExpr 0x1283477d8 '<dependent type>'
|-UnresolvedLookupExpr 0x128347538 '<overloaded function type>' lvalue (no ADL) = 'copy' 0x12824b430
|-CallExpr 0x128347620 '<dependent type>'
| |-UnresolvedLookupExpr 0x1283475a0 '<overloaded function type>' lvalue (no ADL) = 'begin' 0x12093f3b8 0x120f878b0 0x120f883f0 0x120f889b0
| `-DeclRefExpr 0x128347600 'const std::array<float, size>':'const array<float, size>' lvalue ParmVar 0x128347158 'input' 'const std::array<float, size> &'
|-CallExpr 0x1283476e8 '<dependent type>'
| |-UnresolvedLookupExpr 0x128347668 '<overloaded function type>' lvalue (no ADL) = 'end' 0x12093f8d8 0x120f87e70 0x120f88f70 0x120f89530
| `-DeclRefExpr 0x1283476c8 'const std::array<float, size>':'const array<float, size>' lvalue ParmVar 0x128347158 'input' 'const std::array<float, size> &'
`-CallExpr 0x1283477b0 '<dependent type>'
  |-UnresolvedLookupExpr 0x128347730 '<overloaded function type>' lvalue (no ADL) = 'begin' 0x12093f3b8 0x120f878b0 0x120f883f0 0x120f889b0
  `-DeclRefExpr 0x128347790 'float[size]' lvalue Var 0x128347458 'output' 'float[size]'
```
So since all the stdlib calls are dependent, they are stuck as `UnresolvedLookupExpr`s and it's not decided which `begin`, `end` and `copy` will eventually be instantiated (e.g. could be specializations), so IWYU does not have the information necessary.

**Comment by carlosgalvezp** [comment]
Thank you for the clarification! I suppose that explains also why sometimes we observe some "flaky" behavior i.e that IWYU can't decide if it wants and include or not. This could happen in the following scenario:

- foo.h              (defines template function)
- foo.cpp          (#includes foo.h, does not instantiate the template)
- foo_test.cpp  (#includes foo.h, instantiates the template)

If foo.cpp does _not_ instantiate the template, then we end up in this scenario and IWYU wants to remove the header; but if foo_test.cpp does indeed instantiate it, then it determines that the include is wanted.

**Comment by carlosgalvezp** [comment]
A more realistic example of this use case is as follows:

```cpp
$ cat foo.h
#pragma once

#include <algorithm>
#include <array>
#include <cstddef>
#include <iterator>

template <std::size_t size>
void foo(std::array<float, size> const& input)
{
    float output[size]{};
    std::copy(std::begin(input), std::end(input), std::begin(output));
}
```
```
$ cat foo.cpp
#include "foo.h"

int main()
{
  std::array<float, 10> x{};
  foo(x);
}
```

So `foo.cpp` is actually instantiating the template with a concrete type. Unfortunately I still see `UnresolvedLookupExpr` in the logs. How come this happens, if it's resolved?

**Comment by kimgr** [comment]
Interesting. There's probably more to this than I've understood so far.

**Comment by kimgr** [comment]
Fixed on master as of cad1a68c768c2b27ec7bba37e3e9f39456b7bc3e
